### PR TITLE
Make read mapping conditional.

### LINKED ISSUE #834: Check unconditional read mapping in initializeData
`initializeData()` contains an unconditional write mapping. Why?

https://github.com/precice/precice/blob/98b09c8fd184726530c33b23d6400d4ff1cf6d02/src/precice/impl/SolverInterfaceImpl.cpp#L336-L338

In `advance` we have such a condition

https://github.com/precice/precice/blob/a53645cc7bc2ad0a9abba51a1aecbca64aa4022a/src/precice/impl/SolverInterfaceImpl.cpp#L411-L415

Connected to this is the question whether `initializeData` should be called even if no data needs to be initialized (we currently explain it that way in the adapter example)? If yes, how do we treat the mappings there then?

Related to #824

### PR DESCRIPTION
**Pull request by BenjaminRodenberg**
## Main changes of this PR

Closes #834.

Expose `CouplingScheme::sendsInitializedData` to the public.
Check for `sendsInitializedData` to determine whether mapping is necessary.

## Motivation and additional information

See #834.

## Author's checklist

* [x] ~~I added a changelog file with `make changelog` if there are user-observable changes since the last release.~~ Not visible for users.
* [x] ~~I added a test to cover the proposed changes in our test suite.~~ I updated an existing test.
* [x] I ran `make format` to ensure everything is formatted correctly.
* [x] I sticked to C++14 features.
* [x] I sticked to CMake version 3.16.3.
* [x] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [ ] Does the changelog entry make sense? Is it formatted correctly?
* [ ] Do you understand the code changes?

### CODE REVIEW

File: tests/serial/action-timings/ActionTimingsExplicit.cpp [hunk 1]
```diff
@@ -62,8 +62,8 @@ BOOST_AUTO_TEST_CASE(ActionTimingsExplicit)
 
   if (context.isNamed("SolverOne")) {
     BOOST_TEST(action::RecorderAction::records.size() == 2);
-    BOOST_TEST(action::RecorderAction::records.at(0).timing == action::Action::WRITE_MAPPING_PRIOR);
-    BOOST_TEST(action::RecorderAction::records.at(1).timing == action::Action::WRITE_MAPPING_POST);
+    BOOST_TEST(action::RecorderAction::records.at(0).timing == action::Action::READ_MAPPING_PRIOR);
+    BOOST_TEST(action::RecorderAction::records.at(1).timing == action::Action::READ_MAPPING_POST);
```
Comment (BenjaminRodenberg) [inline tests/serial/action-timings/ActionTimingsExplicit.cpp:66]: I added `initialize="True"` in this test to actually check whether the branch is entered. The first participant then only reads the initial data. It does now write data at this point.

File: tests/serial/action-timings/ActionTimingsImplicit.cpp [hunk 2]
```diff
@@ -63,21 +63,15 @@ BOOST_AUTO_TEST_CASE(ActionTimingsImplicit)
 
   dt = interface.initialize();
   BOOST_TEST(dt == 1.0);
-
   if (context.isNamed("SolverOne")) {
-    BOOST_TEST(action::RecorderAction::records.size() == 2);
-    BOOST_TEST(action::RecorderAction::records.at(0).timing == action::Action::WRITE_MAPPING_PRIOR);
-    BOOST_TEST(action::RecorderAction::records.at(1).timing == action::Action::WRITE_MAPPING_POST);
+    BOOST_TEST(action::RecorderAction::records.empty());
```
Comment (BenjaminRodenberg) [inline tests/serial/action-timings/ActionTimingsImplicit.cpp:67]: Without initialization, Nothing actually happens here for the first participant.

File: tests/serial/action-timings/ActionTimingsImplicit.cpp [hunk 3]
```diff
@@ -63,21 +63,15 @@ BOOST_AUTO_TEST_CASE(ActionTimingsImplicit)
...
   } else {
     BOOST_TEST(context.isNamed("SolverTwo"));
-    BOOST_TEST(action::RecorderAction::records.size() == 4);
-    BOOST_TEST(action::RecorderAction::records.at(0).timing == action::Action::WRITE_MAPPING_PRIOR);
-    BOOST_TEST(action::RecorderAction::records.at(1).timing == action::Action::WRITE_MAPPING_POST);
-    BOOST_TEST(action::RecorderAction::records.at(2).timing == action::Action::READ_MAPPING_PRIOR);
-    BOOST_TEST(action::RecorderAction::records.at(3).timing == action::Action::READ_MAPPING_POST);
+    BOOST_TEST(action::RecorderAction::records.size() == 2);
+    BOOST_TEST(action::RecorderAction::records.at(0).timing == action::Action::READ_MAPPING_PRIOR);
+    BOOST_TEST(action::RecorderAction::records.at(1).timing == action::Action::READ_MAPPING_POST);
```
Comment (BenjaminRodenberg) [inline tests/serial/action-timings/ActionTimingsImplicit.cpp:72]: Second participant only reads (again: no intialization.)

File: tests/serial/action-timings/ActionTimingsExplicit.cpp (hunk 1 above)
Comment (uekerman) [reply tests/serial/action-timings/ActionTimingsExplicit.cpp:66]: > It does now write data at this point

It does no**t** write data at this point ??

**Review by uekerman — APPROVED** [review]
Looks good to me.
Probably worth a changelog entry.

File: tests/serial/action-timings/ActionTimingsExplicit.cpp (hunk 1 above)
Comment (BenjaminRodenberg) [reply tests/serial/action-timings/ActionTimingsExplicit.cpp:66]: Yes. Important difference!

**Pull request merged**